packages/flutter/lib/src/widgets/context_menu_controller.dart

```diff
@@ -0,0 +1,132 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'framework.dart';
+import 'inherited_theme.dart';
+import 'navigator.dart';
+import 'overlay.dart';
+
+/// Builds and manages a context menu at a given location.
+///
+/// There can only ever be one context menu shown at a given time in the entire
+/// app.
+///
+/// {@tool dartpad}
+/// This example shows how to use a GestureDetector to show a context menu
+/// anywhere in a widget subtree that receives a right click or long press.
+///
+/// ** See code in examples/api/lib/material/context_menu/context_menu_controller.0.dart **
+/// {@end-tool}
+class ContextMenuController {
+  /// Shows the given context menu.
+  ///
+  /// Since there can only be one shown context menu at a time, calling this
+  /// constructor will also remove any other context menu that is visible.
+  ContextMenuController({
+    this.onRemove,
+    required BuildContext context,
+    required WidgetBuilder contextMenuBuilder,
+    Widget? debugRequiredFor,
+  }) {
+    _show(
```

Comment:
You make some good points.  The interesting thing about ContextMenuController is that it can only ever be showing one instance of a context menu at a time, so I made the API try to reinforce that idea.  But I could refactor things like TextField to take a ContextMenuController instance instead of contextMenuBuilder.  What do you think @LongCatIsLooong?

It would feel better to call `contextMenuController.remove()` instead of `ContextMenuController.removeAny()` in places like [this example](https://github.com/flutter/flutter/pull/107193/files#diff-c37e45d149cf86a6565f1eb32f738d5faaf4170b8bd4f1f6421651f0af2f1bfaR56) at least.  I'll play around with this and see what it looks like.

Reply:
I think it make sense to have removeAny.

I think there are two ways:
1. constructor shows the context menu, and dispose close the context menu.
2. constructor does not show context menu, add an additional `show` API

If we go with (1) we should rename remove/removeAny to dispose/disposeAll. This is kinda weird though because calling this disposeAll will dispose other controller where the caller does not own, it creates a weird ownership model. It is probably better to go with singleton model if we go down this route

If we want to makes the controller a pass in parameter TextField, we can only go with (2).

Reply:
I went with 2 after trying out both.

I also tried passing in the controller to TextField, but I decided against it for now.  The catch here is that in most practical cases that I tried, the contextMenuBuilder needed to be updated for each subsequent call to `show`.  So it wasn't possible to pass the contextMenuBuilder into the constructor and then just call `show` on that instance every time, say, the user right clicks.  Instead, each time the user right clicks, I need to pass in a new contextMenuBuilder to `show`.  That negates the benefit of single long-lived controller instance that the user passes in along with their contextMenuBuilder.

I plan to revisit this when OverlayPortal is released, though I'm not sure how drastically it will change this kind of thing.